pkg/operator/clustermembercontroller/clustermembercontroller.go

```diff
@@ -89,87 +107,586 @@ func (c *ClusterMemberController) reconcileMembers(ctx context.Context, recorder
 	}
 
-	// etcd is healthy, decide if we need to scale
-	podToAdd, err := c.getEtcdPodToAddToMembership(ctx)
-	switch {
-	case err != nil:
-		return fmt.Errorf("could not get etcd pod: %w", err)
-	case podToAdd == nil:
-		// no more work left to do
-		return nil
-	default:
-		recorder.Eventf("FoundPodToScale", "found pod to add to etcd membership: %v", podToAdd.Name)
+	// Set desired cluster size
+	// TODO(haseeb): Always update this to react to control plane size change in cluster config?
+	if c.desiredControlPlaneSize == 0 {
+		desiredControlPlaneSize, err := ceohelpers.GetMastersReplicaCount(ctx, c.kubeClient)
+		if err != nil {
+			return fmt.Errorf("failed to get control-plane replica count: %w", err)
+		}
+		c.desiredControlPlaneSize = desiredControlPlaneSize
+	}
+
+	// Update preDrain hooks on Machines
+	// Blocks or unblocks machine deletion
+	if err = c.reconcileMachineHooks(ctx); err != nil {
+		return fmt.Errorf("failed to reconcile machine hooks: %v", err)
 	}
 
-	etcdHost, err := c.getEtcdPeerHostToScale(podToAdd)
+	// Add a learner member if next peer found
+	peerURL, err := c.getEtcdPeerURLToAdd(ctx)
 	if err != nil {
-		return fmt.Errorf("could not get etcd peer host :%w", err)
+		return fmt.Errorf("could not get etcd peerURL to add :%w", err)
+	}
+	if peerURL != "" {
+		if err = c.etcdClient.MemberAddAsLearner(ctx, peerURL); err != nil {
+			return fmt.Errorf("failed to add learner member: %w", err)
+		}
 	}
-	err = c.etcdClient.MemberAdd(ctx, fmt.Sprintf("https://%s:2380", etcdHost))
+
+	// Scale up the cluser by promoting a learner member
+	// if there is room for more voting members
+	// Either during bootstrap to N members, or vertical scaling
+	if err := c.ensureEtcdLearnerPromotion(ctx, recorder); err != nil {
+		return fmt.Errorf("failed to ensure learner promotion: %w", err)
+	}
+
+	// Get next member to scale-down
+	// This could be for scaling down in a vertical scaling scenario
+	// or to remove a learner member to unblock a machine pending deletion
+	member, err := c.getEtcdMemberToScaleDown(ctx)
 	if err != nil {
-		return fmt.Errorf("could not add member :%w", err)
+		return fmt.Errorf("could not get etcd member for removal:%w", err)
+	}
+	// Scale down if next member found
+	if member != nil {
+		if err = c.etcdClient.MemberRemoveID(ctx, member); err != nil {
+			return fmt.Errorf("failed to remove member: %w", err)
+		}
 	}
+
 	return nil
 }
 
-func (c *ClusterMemberController) getEtcdPodToAddToMembership(ctx context.Context) (*corev1.Pod, error) {
-	// list etcd member pods
-	pods, err := c.podLister.List(labels.Set{"app": "etcd"}.AsSelector())
+// getEtcdPeerURLToAdd returns a PeerURL of the next etcd instance to be added as a learner to the cluster. If
+// no elegible member is found return an empty string.
+func (c *ClusterMemberController) getEtcdPeerURLToAdd(ctx context.Context) (string, error) {
+	members, err := c.etcdClient.MemberList(ctx)
```

Comment:
what's the cost of this?

Note that your controller is driven by e.g. node updates. Nodes have heartbeats and hence you have quite a number of updates in a big cluster.

Reply:
Maybe you can have an informer that is filtered by the master label.